### PR TITLE
Implement a deep-copy functionality and use it for deep-merge

### LINKED ISSUE #10: Always allocate a new value
There is a comment [here](https://github.com/adutra/goalesce/blob/main/coalesce.go#L36):

> // The function never modifies its inputs. It may return one of the (unmodified) inputs as the coalesced value, but
> // whenever changes to the inputs are required, the function always returns an entirely newly-allocated value.

Can I ask that we change the logic here so that a newly allocated value is always returned? I have some concerns about the fact that the behaviour is inconsistent between when a value is changed vs when it isn't. 

It makes it hard to reason about whether a returned value that is then mutated outside the Goalesce logic is actually mutating the original object or not (as this would differ depending on whether the object was changed by Goalesce). 

This class of errors will also be hard to catch in tests.

### PR DESCRIPTION
**Pull request by adutra**
Fixes #10.

Opening a draft for now to share what I currently have. 

I have renamed main `Coalesce` function to `DeepMerge`, to get a bit of symmetry with `DeepCopy`.

If we ever implement in-place, I would see two more functions:

* `DeepCopyInto`
* `DeepMergeInto`

TODOs:

- [x] Tests for DeepCopy
- [x] Tests for DeepMerge verifying that objects were copied
- [x] DeepCopy examples
- [x] DeepCopy examples in README.md
- [x] Improve godocs

### CODE REVIEW

File: pointer_test.go [hunk 1]
```diff
@@ -78,13 +78,13 @@ func Test_mainCoalescer_coalescePointer(t *testing.T) {
 			"cycle 1",
 			simpleCycle(),
 			(*cycle)(nil),
-			simpleCycle(), // unnoticed
+			&cycle{Cycle: &cycle{}},
```
Comment (adutra) [inline pointer_test.go:81]: Cycle detection changed behavior a bit because of deep-copy. But it still works as before, barring the issues raised in #24.

**Comment by adutra** [comment]
@Miles-Garnsey if are you are brave enough :-D this is ready for review. The code coverage remains 100%.

**Comment by adutra** [comment]
Also fixes #21.

File: array.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	if c.arrayMerger != nil {
+		return c.arrayMerger(v1, v2)
+	}
+	return c.deepMergeAtomic(v1, v2)
+}
+
+// deepMergeArrayByIndex is an alternate array merger that merges arrays with by-index semantics. It
+// is not the default merge strategy for arrays; it is only activated if an array merger has been
+// registered through one of the options: WithDefaultArrayMergeByIndex, WithArrayMergeByIndex.
+func (c *coalescer) deepMergeArrayByIndex(v1, v2 reflect.Value) (reflect.Value, error) {
```
Comment (adutra) [inline array.go:38]: This is actually the implementation of #21 .

**Review by adejanovski — APPROVED** [review]

**Pull request merged**